### ISSUE TITLE
shared_mem corrupted after stop and restart of nodeos - DAWN 4.0

### ISSUE THREAD

**Issue by ankh2054** [EOSIO/eos#3047]
OS: Ubuntu 16.04
EOS: Dawn 4.0

Since upgrading to DAWN-4.0 any stop and restart of nodes causes ` shared_mem/ ` to get corrupted.
Tested on 2 different nodes and have confirmed the same issue with 5 other users on Cryptolions testnet.

**settings in config.ini**

```
    get-transactions-time-limit = 3

    allowed-connection = any

    log-level-net-plugin = info
    max-clients = 120
    connection-cleanup-period = 30
    network-version-match = 0
    sync-fetch-span = 2000
    enable-stale-production = false
    required-participation = 33

    mongodb-queue-size = 256

    plugin = eosio::producer_plugin
    plugin = eosio::chain_api_plugin
    plugin = eosio::history_plugin
    plugin = eosio::history_api_plugin
    plugin = eosio::chain_plugin
```

On testnet we utilise following start and stop scripts.

**Start script:**
```
NODEOS=/opt/eos//build/programs/nodeos/nodeos
DATADIR=/opt/JungleTestnet
$DATADIR/stop.sh
$NODEOS --data-dir $DATADIR --config-dir $DATADIR "$@" > $DATADIR/stdout.txt 2> $DATADIR/stderr.txt &  echo $! > $DATADIR/nodeos.pid
```

**Stop script:**
```
DIR=/opt/JungleTestnet

    if [ -f $DIR"/nodeos.pid" ]; then
        pid=$(cat $DIR"/nodeos.pid")
        echo $pid
        kill $pid
        rm -r $DIR"/nodeos.pid"

        echo -ne "Stoping Nodeos"

        while true; do
            [ ! -d "/proc/$pid/fd" ] && break
            echo -ne "."
            sleep 1
        done
        echo -ne "\rNodeos stopped. \n"

    fi
```

**ISSUE**
Stop nodeos.(using kill -15 ) OR using stop script and restart nodeos, the following error occurs.

```
2906719ms thread-0   chain_plugin.cpp:125          plugin_initialize    ] initializing chain plugin
2906721ms thread-0   block_log.cpp:120             open                 ] Log is nonempty
2906721ms thread-0   block_log.cpp:123             open                 ] my->head->block_num(): 321695 
2906721ms thread-0   block_log.cpp:129             open                 ] Index is nonempty
2907046ms thread-0   main.cpp:99                   main                 ] 10 assert_exception: Assert Exception
size.value*sizeof(T) < MAX_ARRAY_ALLOC_SIZE: 
    {}
    thread-0  raw.hpp:560 unpack
```

**Resolution**

Deleting shared_mem/ and restarting nodeos gets around the error.

**Comment by bytemaster** [comment]
I am unable to reproduce this error.  I do not think this is from the sharedmem but from the fork database being loaded as forkdb isn the one that would attempt a raw unpack.

**Comment by bytemaster** [comment]
Here is my console output from "kill through restart".  
```
2372503ms thread-0   producer_plugin.cpp:587       block_production_loo ] Produced block 00001642b2cada56... #5698 @ 2018-05-14T18:39:32.500 signed by eosio [trxs: 0, lib: 5698, confirmed: 0]
2372718ms thread-0   net_plugin.cpp:2940           plugin_shutdown      ] shutdown..
2372718ms thread-0   net_plugin.cpp:2943           plugin_shutdown      ] close acceptor
2372718ms thread-0   net_plugin.cpp:2946           plugin_shutdown      ] close 0 connections
2372718ms thread-0   net_plugin.cpp:2954           plugin_shutdown      ] exit shutdown
2372718ms thread-0   fork_database.cpp:94          close                ] states.size(): 1
2372718ms thread-0   controller.cpp:194            ~controller_impl     ] db.revision(): 5698 head->block_num: 5698 blog.read_head()->block_num(): 5698
[dlarimer@danpro ~/eos/programs/nodeos]$ ./nodeos -d dawn4 --producer eosio -e  --plugin eosio::chain_api_plugin --plugin eosio::history_api_plugin
2377825ms thread-0   chain_plugin.cpp:125          plugin_initialize    ] initializing chain plugin
2377827ms thread-0   block_log.cpp:120             open                 ] Log is nonempty
2377827ms thread-0   block_log.cpp:123             open                 ] my->head->block_num(): 5698
2377827ms thread-0   block_log.cpp:129             open                 ] Index is nonempty
2377843ms thread-0   http_plugin.cpp:247           plugin_initialize    ] configured http to listen on 127.0.0.1:8888
2377843ms thread-0   net_plugin.cpp:2808           plugin_initialize    ] Initialize net plugin
2377843ms thread-0   net_plugin.cpp:2829           plugin_initialize    ] host: 0.0.0.0 port: 9876
2377843ms thread-0   net_plugin.cpp:2901           plugin_initialize    ] my node_id is c65c09d5c29be9cb46d1d02f43034dd8970c54e60dc4fb96d067975e32eab66f
2377844ms thread-0   main.cpp:94                   main                 ] nodeos version dacc1b09
2377844ms thread-0   main.cpp:95                   main                 ] eosio root is /Users/dlarimer/Library/Application Support
2377844ms thread-0   chain_plugin.cpp:253          plugin_startup       ] starting chain in read/write mode
2377844ms thread-0   chain_plugin.cpp:258          plugin_startup       ] Blockchain started; head block is #5698, genesis timestamp is 2018-03-01T12:00:00.000
2377844ms thread-0   http_plugin.cpp:285           plugin_startup       ] start listening for http requests
2377920ms thread-0   chain_api_plugin.cpp:62       plugin_startup       ] starting chain_api_plugin
2377921ms thread-0   http_plugin.cpp:325           add_handler          ] add api url: /v1/chain/abi_bin_to_json
2377921ms thread-0   http_plugin.cpp:325           add_handler          ] add api url: /v1/chain/abi_json_to_bin
2377921ms thread-0   http_plugin.cpp:325           add_handler          ] add api url: /v1/chain/get_account
2377921ms thread-0   http_plugin.cpp:325           add_handler          ] add api url: /v1/chain/get_block
2377921ms thread-0   http_plugin.cpp:325           add_handler          ] add api url: /v1/chain/get_code
2377921ms thread-0   http_plugin.cpp:325           add_handler          ] add api url: /v1/chain/get_currency_balance
2377921ms thread-0   http_plugin.cpp:325           add_handler          ] add api url: /v1/chain/get_currency_stats
2377921ms thread-0   http_plugin.cpp:325           add_handler          ] add api url: /v1/chain/get_info
2377921ms thread-0   http_plugin.cpp:325           add_handler          ] add api url: /v1/chain/get_required_keys
2377921ms thread-0   http_plugin.cpp:325           add_handler          ] add api url: /v1/chain/get_table_rows
2377921ms thread-0   http_plugin.cpp:325           add_handler          ] add api url: /v1/chain/push_block
2377921ms thread-0   http_plugin.cpp:325           add_handler          ] add api url: /v1/chain/push_transaction
2377921ms thread-0   http_plugin.cpp:325           add_handler          ] add api url: /v1/chain/push_transactions
2377921ms thread-0   history_api_plugin.cpp:44     plugin_startup       ] starting history_api_plugin
2377922ms thread-0   http_plugin.cpp:325           add_handler          ] add api url: /v1/history/get_actions
2377922ms thread-0   http_plugin.cpp:325           add_handler          ] add api url: /v1/history/get_transaction
2377922ms thread-0   producer_plugin.cpp:383       plugin_startup       ] producer plugin:  plugin_startup() begin
2377922ms thread-0   producer_plugin.cpp:390       plugin_startup       ] Launching block production for 1 producers.
2377945ms thread-0   producer_plugin.cpp:403       plugin_startup       ] producer plugin:  plugin_startup() end
2377945ms thread-0   net_plugin.cpp:2913           plugin_startup       ] starting listener, max clients is 25
2378005ms thread-0   producer_plugin.cpp:587       block_production_loo ] Produced block 00001643c9b36115... #5699 @ 2018-05-14T18:39:38.000 signed by eosio [trxs: 0, lib: 5699, confirmed: 0]
```

**Comment by stvenyin** [comment]
See.

**Comment by winlin** [comment]
we still suffer this error in dawn-v4.1.0

**Comment by bitcodernull** [comment]
I have met the same errors

**Comment by jorfermo** [comment]
Having same error with regular stop + start in all my instances:

617226ms thread-0   chain_plugin.cpp:130          plugin_initialize    ] initializing chain plugin
617226ms thread-0   block_log.cpp:120             open                 ] Log is nonempty
617226ms thread-0   block_log.cpp:123             open                 ] my->head->block_num(): 194289
617226ms thread-0   block_log.cpp:129             open                 ] Index is nonempty
617455ms thread-0   main.cpp:106                  main                 ] 10 assert_exception: Assert Exception
size.value*sizeof(T) < MAX_ARRAY_ALLOC_SIZE:
    {}
    thread-0  raw.hpp:560 unpack

Only resync fixes it

**Comment by arhag** [comment]
Resolved by #3392.
